src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -425,6 +425,7 @@ private bool IsMinimalDelta
 
         // progress:
         private bool _tableIndicesAreComplete;
+        private bool _sourceDebugDocumentsAdded;
```

Comment:
Could you please explain why we need this flag?

Reply:
The reason why is we need to hook the `AddDocument` call and insert the real names for non-source files. This flag is set once we'ev ensured all the source files are in the debug document table. From then on the calls to `AddDocument` will pull from the non-source file names

Reply:
Could the enumerator be nullable instead of having separate bool?